container_pipeline/cleanup_registry/garbagecollector.py

```diff
@@ -10,53 +12,118 @@ class GarbageCollector(object):
         self._index_git = index_git
+        self._gc_exceptions = []
+        self._gc_match_only = []
         # Formulate the registry url
         self._registry_url = str.format("{schema}://{host}{port}/v2",
                                         schema="https" if registry_secure else "http",
                                         host=registry_host,
                                         port="" if not registry_port else ":" + registry_port
                                         )
-        self._index_location = self._index_location + "/index.d"
+        self._index_d_location = self._index_location + "/index.d"
         # Setup reg info object to query the registry and cache metadata.
+        self._index_containers = {}
+        self._mismatched = {}
+
+    def _query_registry(self):
         if self._verbose:
             print("Gathering registry metadata...")
         self._registry_info = lib.RegistryInfo(self._registry_url)
-        self.index_containers = {}
-        self.mismatched = {}
 
-    def _delete_mismatched(self):
+    def _query_index(self):
+        """
+        Query a cloned or local container index.
+        :return:
+        """
+        index_files = glob(self._index_d_location + "/*.yml")
+        for index_file in index_files:
+            if "index_template" not in index_file:
+                data = lib.load_yaml(index_file)
+                if "Projects" not in data:
+                    raise Exception("Invalid index file")
+                for entry in data["Projects"]:
+                    app_id = entry["app-id"]
+                    job_id = entry["job-id"]
+                    desired_tag = entry["desired-tag"]
+                    container_name = str.format(
+                        "{namespace}{name}",
+                        namespace=(str(app_id) + "/") if str(app_id) != "library" else "",
+                        name=str(job_id)
+                    )
+                    container_tag = container_name + ":" + str(desired_tag)
+                    if container_name not in self._index_containers:
+                        #
+                        self._index_containers[container_name] = []
+                        if not BuildTracker(container_tag).is_running():
+                            self._index_containers[container_name].append(desired_tag)
+
+    def _prep_index(self):
+        """
+        Prepares the container index, and queries the same.
+        :return:
+        """
+        index_preped = False
+        if self._local_index:
+            # If local index is set, check if index files are present at expected location.
+            if not lib.path.exists(self._index_location):
+                raise Exception("Local index specified, but does not exist")
+            index_preped = True
+        elif self._index_git:
+            # Otherwise check if index git is specified, if so clone it.
+            if self._verbose:
+                print("Cloning container index...")
```

Comment:
a better implementation of print based on verbose flag could be

```
def _print(msg, verbose=True):
   if verbose:
           print msg
```
and have all the print statements replaced via `_print`
where in, for mandatory print occurrences you can just
```
_print("I must print on stdout.")
```
or for optional (if verbose), you can
```
_print("I could print if you asked.", versbose=verbose_falg_value)
```